packages/promotion/src/repositories/promotion-rule.ts

```diff
@@ -0,0 +1,124 @@
+import { Context, DAL } from "@medusajs/types"
+import { DALUtils, MedusaError, arrayDifference } from "@medusajs/utils"
+import {
+  FilterQuery as MikroFilterQuery,
+  FindOptions as MikroOptions,
+} from "@mikro-orm/core"
+
+import { SqlEntityManager } from "@mikro-orm/postgresql"
+import { PromotionRule } from "@models"
+import { CreatePromotionRuleDTO, UpdatePromotionRuleDTO } from "@types"
+
+export class PromotionRuleRepository extends DALUtils.MikroOrmBaseRepository {
+  protected readonly manager_: SqlEntityManager
+
+  constructor({ manager }: { manager: SqlEntityManager }) {
+    // @ts-ignore
+    // eslint-disable-next-line prefer-rest-params
+    super(...arguments)
+    this.manager_ = manager
+  }
+
+  async find(
+    findOptions: DAL.FindOptions<PromotionRule> = { where: {} },
+    context: Context = {}
+  ): Promise<PromotionRule[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+
+    const findOptions_ = { ...findOptions }
+    findOptions_.options ??= {}
+
+    return await manager.find(
+      PromotionRule,
+      findOptions_.where as MikroFilterQuery<PromotionRule>,
+      findOptions_.options as MikroOptions<PromotionRule>
+    )
+  }
+
+  async findAndCount(
+    findOptions: DAL.FindOptions<PromotionRule> = { where: {} },
+    context: Context = {}
+  ): Promise<[PromotionRule[], number]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+
+    const findOptions_ = { ...findOptions }
+    findOptions_.options ??= {}
+
+    return await manager.findAndCount(
+      PromotionRule,
+      findOptions_.where as MikroFilterQuery<PromotionRule>,
+      findOptions_.options as MikroOptions<PromotionRule>
+    )
+  }
+
+  async delete(ids: string[], context: Context = {}): Promise<void> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+    await manager.nativeDelete(PromotionRule, { id: { $in: ids } }, {})
+  }
+
+  async create(
+    data: CreatePromotionRuleDTO[],
+    context: Context = {}
+  ): Promise<PromotionRule[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+
+    const promotionRule = data.map((promotionRule) => {
+      return manager.create(PromotionRule, promotionRule)
+    })
+
+    manager.persist(promotionRule)
+
+    return promotionRule
+  }
+
+  async update(
+    data: UpdatePromotionRuleDTO[],
+    context: Context = {}
+  ): Promise<PromotionRule[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+    const promotionRuleIds = data.map((promotionRule) => promotionRule.id)
+    const existingPromotionRules = await this.find(
+      {
+        where: {
+          id: {
+            $in: promotionRuleIds,
+          },
+        },
+      },
+      context
+    )
+
+    const dataAndExistingIdDifference = arrayDifference(
+      data.map((d) => d.id),
+      existingPromotionRules.map((plr) => plr.id)
+    )
+
+    if (dataAndExistingIdDifference.length) {
```

Comment:
Agreed, I've moved this to the services in the update ops branch.

Reply:
nice :) could you create a ticket in linear to check all existing modules and align them if needed :)

Reply:
@riqwan, where did you add these changes? I have a similar comment on my PR, and I was curious to see what approach you took.

Reply:
@olivermrbl @adrien2p has played around with this here - https://github.com/medusajs/medusa/compare/chore/abstract-module-repository?expand=1

The same could be done with services and this logic could be pushed there. Its a more explicit version than the config approach, but I think some level of explicitness is good to have, atleast for now. This def improves from what we have atm

Reply:
the services are played around with too 😄  - https://github.com/medusajs/medusa/compare/chore/abstract-module-service?expand=1

Reply:
This is pretty fantastic. I think this will make us move significantly faster for other modules.

I am not sure about the DX (as I also told Adrien), but we don't have to worry about that too much until users will need to build their own modules. Right now, it's more important to just reduce the amount of boilerplate code on our side.

Reply:
@olivermrbl should we then proceed with opening those pr's or do you prefer to wait a bit before?